src/hotspot/share/jfr/recorder/stacktrace/jfrAsyncStackTrace.cpp

```diff
@@ -0,0 +1,144 @@
+#include "jfr/recorder/repository/jfrChunkWriter.hpp"
+#include "jfr/recorder/storage/jfrBuffer.hpp"
+#include "runtime/threadCrashProtection.hpp"
+#include "runtime/vframe.inline.hpp"
+
+JfrAsyncStackFrame::JfrAsyncStackFrame(const Method* method, int bci, u1 type, int lineno, const InstanceKlass* ik) :
+  _klass(ik), _method(method), _line(lineno), _bci(bci), _type(type) {}
+
+
+JfrAsyncStackTrace::JfrAsyncStackTrace(JfrAsyncStackFrame* frames, u4 max_frames) :
+  _frames(frames),
+  _nr_of_frames(0),
+  _max_frames(max_frames),
+  _reached_root(false)
+  {}
+
+bool JfrAsyncStackTrace::record_async(JavaThread* jt, const frame& frame) {
+  NoHandleMark nhm;
+
+  assert(jt != nullptr, "invariant");
+  Thread* current_thread = Thread::current_or_null_safe();
+  if (current_thread == nullptr) {
+    return false;
+  }
+  assert(current_thread->in_asgct(), "invariant");
+
+  u4 count = 0;
+  _reached_root = true;
+
+  JfrVframeStream vfs(jt, frame, false, true);
+
+  while (!vfs.at_end()) {
+    if (count >= _max_frames) {
+      _reached_root = false;
+      break;
+    }
+    const Method* method = vfs.method();
+    if (method == nullptr || !Method::is_valid_method(method)) {
+      // we throw away everything we've gathered in this sample since
+      // none of it is safe
+      return false;
+    }
+    u1 type = vfs.is_interpreted_frame() ? JfrStackFrame::FRAME_INTERPRETER : JfrStackFrame::FRAME_JIT;
+    int bci = 0;
+    if (method->is_native()) {
+      type = JfrStackFrame::FRAME_NATIVE;
+    } else {
+      bci = vfs.bci();
+    }
+
+    intptr_t* frame_id = vfs.frame_id();
+    vfs.next_vframe();
+    if (type == JfrStackFrame::FRAME_JIT && !vfs.at_end() && frame_id == vfs.frame_id()) {
+      // This frame and the caller frame are both the same physical
+      // frame, so this frame is inlined into the caller.
+      type = JfrStackFrame::FRAME_INLINE;
+    }
+    _frames[count] = JfrAsyncStackFrame(method, bci, type, method->line_number_from_bci(bci), method->method_holder());
+    count++;
+  }
+  _nr_of_frames = count;
+  return count > 0;
+}
+
+class JfrAsyncStackTraceStoreCallback : public CrashProtectionCallback {
+ public:
+  JfrAsyncStackTraceStoreCallback(const JfrAsyncStackTrace* asyncTrace, JfrStackTrace* trace) :
+  _asyncTrace(asyncTrace), _trace(trace), _success(false) {}
+  virtual void call() {
+    _success = _asyncTrace->inner_store(_trace);
+  }
+  bool success() { return _success; }
+
+ private:
+  const JfrAsyncStackTrace* _asyncTrace;
+  JfrStackTrace* _trace;
+  bool _success;
+};
+
+bool JfrAsyncStackTrace::inner_store(JfrStackTrace* trace) const {
+  traceid hash = 1;
+  for (u4 i = 0; i < _nr_of_frames; i++) {
+    const JfrAsyncStackFrame& frame = _frames[i];
+    if (!Method::is_valid_method(frame._method)) {
+      // we throw away everything we've gathered in this sample since
+      // none of it is safe
+      return false;
+    }
+    const traceid mid = JfrTraceId::load(frame._method);
```

Comment:
This is JfrTraceId::load() barrier code:

inline traceid JfrTraceIdLoadBarrier::load(const Klass* klass, const Method* method) {
   assert(klass != nullptr, "invariant");
   assert(method != nullptr, "invariant");
   if (should_tag(method)) {
     SET_METHOD_AND_CLASS_USED_THIS_EPOCH(klass);
     SET_METHOD_FLAG_USED_THIS_EPOCH(method);
     assert(METHOD_AND_CLASS_USED_THIS_EPOCH(klass), "invariant");
     assert(METHOD_FLAG_USED_THIS_EPOCH(method), "invariant");
     enqueue(klass);
     JfrTraceIdEpoch::set_changed_tag_state();
   }
   return (METHOD_ID(klass, method));
}

What happens if the sampler crashes at each of these instructions?

1. Sets bits in the Klass, but not in the Method, and does not enqueue.
2. Set bits in the Klass and Method, but does not enqueue?
3. Sets bits in Klass, Method and enqueues, but does not notify?

Reply:
I removed the crash protection that surrounds the code. It isn't needed anymore.